src/renderer/wg_engine/tvgWgGeometry.cpp

```diff
@@ -456,13 +462,18 @@ void WgGeometryData::appendStroke(const WgPolyline* polyline, const RenderStroke
                     if (stroke->join == StrokeJoin::Bevel) {
                         appendRect(v1 - offset0, v1 + offset0, v1 - offset1, v1 + offset1);
                     } else if (stroke->join == StrokeJoin::Miter) {
-                        WgPoint nrm = (nrm0 + nrm1).normal();
-                        float cosine = nrm.dot(nrm0);
-                        if ((cosine != 0.0f) && (abs(cosine) < 1.0f) && (abs(wdt / cosine) <= stroke->miterlimit * 2)) {
-                            appendRect(v1 + nrm * (wdt / cosine), v1 + offset0, v1 + offset1, v1);
-                            appendRect(v1 - nrm * (wdt / cosine), v1 - offset0, v1 - offset1, v1);
-                        } else {
-                            appendRect(v1 - offset0, v1 + offset0, v1 - offset1, v1 + offset1);
+                        WgPoint nrm = (nrm0 + nrm1);
+                        if(!mathZero(nrm.length())) {
+                            nrm.normalize();
+                            float cosine = nrm.dot(nrm0);
+                            float angle = std::acos(dir0.dot(dir1.negative()));
+                            float miterRatio = 1.0f / (std::sin(angle) * 0.5);
+                            if ((abs(cosine) < 1.0f) && (miterRatio <= stroke->miterlimit)) {
```

Comment:
as I said, I don;t think case when cosine is slightly more than zero is a problem. if it is, there is some underlying cause, not the value of cosine itself. and we should take care about the underlaying case, not the effect.
here one problem I see is division by zero, which should not take place, no matter whether we use this value later or not.